src/main/java/edu/uw/edm/docfinity/config/AppConfig.java

```
@@ -0,0 +1,159 @@
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
+import org.springframework.util.StringUtils;
+import org.springframework.web.client.RestTemplate;
+
+/** @author Jay Shen */
+@Configuration
+public class AppConfig {
+    private KeyManager[] getKeyManagers(final AppProperties appProperties)
+            throws KeyStoreException, FileNotFoundException, IOException, CertificateException,
+                    UnrecoverableKeyException, NoSuchAlgorithmException {
+
+        if (StringUtils.isEmpty(appProperties.getKeystoreLocation())) {
+            return null;
+        }
+
+        // keystore
+        final String keystoreType = "JKS";
+        KeyStore ks = KeyStore.getInstance(keystoreType);
+
+        FileInputStream fis = new FileInputStream(appProperties.getKeystoreLocation());
+        char[] password =
+                appProperties.getKeystorePassword() != null
+                        ? appProperties.getKeystorePassword().toCharArray()
+                        : "changeit".toCharArray();
+        try {
+            ks.load(fis, password);
+        } finally {
+            if (fis != null) {
+                fis.close();
+            }
+        }
+
+        KeyManagerFactory kmf = KeyManagerFactory.getInstance("SunX509");
+        kmf.init(ks, password);
+        return kmf.getKeyManagers();
+    }
+
+    @Bean
+    public CloseableHttpClient httpClient(final AppProperties appProperties) throws Exception {
+        // RequestConfig
+        RequestConfig requestConfig = RequestConfig.custom().setExpectContinueEnabled(true).build();
+
+        // HostnameVerifier
+        HostnameVerifier hostnameVerifier;
+        if (appProperties.isIgnoreHostname()) {
+            hostnameVerifier = new NoopHostnameVerifier();
+        } else {
+            hostnameVerifier = new DefaultHostnameVerifier();
+        }
+
+        // keyManagers
+        KeyManager[] keyManagers = getKeyManagers(appProperties);
```

getKeyManagers can return null, does that need to be handled here?

---

No. It is passed to SSLContext.init and null is a legitimate value. See 
https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLContext.html for detail.

---

What happens if the paremeter is null? Is it expected to work or is it going to fail down the line with some cryptic error message? If it is required for the regular operations maybe we should throw a descriptive error.

---

I spent hours to look into this and to test different cases. Here is what I found:
* keyManagers is null for unit tests, unless we create a fake, valid java keystore for unit tests. I don't like the complication of creating such a fake java keystore just for unit tests. 
* When the keyManagers was null, it was treated the same as no client cert and PWS returned an error response.

I don't think we need to worry about null values, because all our other APIs are handling it the same way and we have not experienced any issues for many years. This does not mean it would never happen, but we don't need to complicate our code to handle such an unlikely case.  What is much more likely could be an expired cert (I hope you are not going to ask me to check the cert validity at this point)?

---

If the error message from PWS is descriptive as you said we should be ok.